pkg/k8sutils/pod.go

```diff
@@ -3,43 +3,125 @@ package k8sutils
 		LabelSelector: "job-name=" + jobName,
 	})
 	if err != nil {
 		return "", err
 	}
 
-	logs := ""
+	var logs strings.Builder
 
 	for _, pod := range list.Items {
 
-		req := k8s.clientset.CoreV1().Pods(namespace).GetLogs(pod.Name, &podLogOpts)
-		podLogs, err := req.Stream(context.TODO())
-		if err != nil {
-			return "", err
+		// Query all logs from containers that have terminated and therefore already had the chance to
+		// produce logs, otherwise the k8s api will return an error
+		for _, container := range getTerminatedContainersWithStatusOfPod(pod) {
+
+			// Query logs of the current selected container
+			logsOfContainer, err := getLogsOfContainer(k8s, pod, namespace, container.name)
+			if err != nil {
+				return "", err
+			}
+
+			// Prepend the container name at the beginning, so we are able to separate logs of different containers
+			// and display a termination error at the beginning, may be more interesting than the logs of the container
+			if container.status.Reason != "Completed" {
+				logs.WriteString("Container ")
+				logs.WriteString(container.name)
+				logs.WriteString(" terminated with an error (Reason: ")
+				logs.WriteString(container.status.Reason)
+
+				// Sometimes the message is not given, to provide prettier logs we just don't print the
+				// message part if it doesn't exist
+				if container.status.Message != "" {
+					logs.WriteString(", Message: ")
+					logs.WriteString(container.status.Message)
+					logs.WriteString(")")
+				} else {
+					logs.WriteString(", ExitCode: ")
+					logs.WriteString(fmt.Sprintf("%d", container.status.ExitCode))
+					logs.WriteString(")")
+				}
+
+				logs.WriteString(":\n")
+			} else {
+				logs.WriteString("Container ")
+				logs.WriteString(container.name)
+				logs.WriteString(":\n")
+			}
+
+			// Finally, append the actual logs of the container or a default message to the log
+			if logsOfContainer != "" {
+				logs.WriteString(logsOfContainer)
+				logs.WriteString("\n")
+			} else {
+				logs.WriteString("<no logs available for container>\n")
+			}
+
+			logs.WriteString("\n")
+		}
+	}
+
+	return logs.String(), nil
+}
+
+type containerStatus struct {
+	name   string
+	status *v1.ContainerStateTerminated
+}
+
+func getLogsOfContainer(k8s *k8sImpl, pod v1.Pod, namespace string, container string) (string, error) {
+	req := k8s.clientset.CoreV1().Pods(namespace).GetLogs(pod.Name, &v1.PodLogOptions{
+		Container: container,
+	})
+
+	podLogs, err := req.Stream(context.TODO())
+	if err != nil {
+		return "", err
+	}
+
+	defer podLogs.Close()
+
+	buf := new(bytes.Buffer)
+	_, err = io.Copy(buf, podLogs)
+	if err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+func getTerminatedContainersWithStatusOfPod(pod v1.Pod) []containerStatus {
+	var containerStatusList []containerStatus
+
+	for index, initContainer := range pod.Spec.InitContainers {
+		if pod.Status.InitContainerStatuses[index].State.Terminated != nil {
```

Comment:
Can we also look at "pending" containers, containers with "Image Pull Back Off", etc...? Or containers that have never properly started? (same question for the line below with `pod.Spec.Containers`)

Reply:
I can change the implementation such that it gathers the status of terminated and pending containers, but since we only call this function to gather the logs of jobs that have been marked as finished, no other container states should be possible. I think that the status of pending containers ("Image Pull Backoff") will only be visible in the events.